File: crates/harness-server/src/comments_routes.rs
//! REST routes for [`Comment`](harness_project::Comment) — Phase 3.8
//! per-requirement discussion threads.
//!
//! Mounted only when [`AppState::comments`] is configured. All
//! endpoints return `503 Service Unavailable` otherwise so callers
//! can distinguish "feature off" from "really broken".
//!
//! Endpoints:
//!
//! - `GET    /v1/requirements/:id/comments`           — list, oldest-first
//! - `POST   /v1/requirements/:id/comments`           — post a top-level comment or reply
//! - `PATCH  /v1/comments/:comment_id`                — edit body
//! - `DELETE /v1/comments/:comment_id`                — delete (cascades to replies)
//!
//! Authoring policy is intentionally minimal in v0:
//!
//! * Author defaults to [`ActivityActor::Human`]. Future work that
//!   wires up real user identity will flip this; agent-authored
//!   comments come from agent-tool callsites that bypass these
//!   routes (no `requirement.comment` tool today).
//! * Trim + non-empty + length-cap on body. Format-agnostic — the
//!   web UI renders markdown via its existing component which owns
//!   sanitisation.
//! * Edits go through the model's [`Comment::edit`] helper so
//!   `updated_at` is bumped consistently with the in-process call
//!   path used by future agent tools.
//!
//! Every successful mutation also emits an
//! [`ActivityKind::Comment`] row into [`ActivityStore`] so the
//! per-Requirement timeline (the `GET /v1/requirements/:id/activities`
//! endpoint) shows the discussion alongside status flips and run
//! lifecycle events. Activity emission is best-effort — a missing
//! `ActivityStore` is silently skipped, mirroring existing patterns
//! in `requirements_routes::record_activity`.

use std::sync::Arc;

use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::{IntoResponse, Json, Response},
    routing::{get, patch},
    Router,
};
use harness_project::{Activity, ActivityActor, ActivityKind, Comment, CommentStore};
use serde::Deserialize;
use serde_json::{json, Value};
use tracing::{error, warn};

use crate::state::AppState;
use crate::state_layers::ProjectLayer;

/// Soft cap on a single comment body. Mirrors typical chat-app
/// limits (Linear: 32 KiB, GitHub: 64 KiB). Past this the model
/// stops being useful as discussion and the agent token budget
/// suffers — push long content into the description / a doc.
const MAX_COMMENT_BODY_BYTES: usize = 32 * 1024;

pub(crate) fn router() -> Router<AppState> {
    Router::new()
        .route(
            "/v1/requirements/:id/comments",
            get(list_comments).post(post_comment),
        )
        .route(
            "/v1/comments/:comment_id",
            patch(edit_comment).delete(delete_comment),
        )
}

#[allow(clippy::result_large_err)]
fn require_comment_store(project: &ProjectLayer) -> Result<Arc<dyn CommentStore>, Response> {
    project.comments.clone().ok_or_else(|| {
        (
            StatusCode::SERVICE_UNAVAILABLE,
            Json(json!({ "error": "comment store not configured" })),
        )
            .into_response()
    })
}

fn internal_error(e: impl std::fmt::Display) -> Response {
    error!(error = %e, "comments error");
    (
        StatusCode::INTERNAL_SERVER_ERROR,
        Json(json!({ "error": e.to_string() })),
    )
        .into_response()
}

fn bad_request(msg: impl Into<String>) -> Response {
    (
        StatusCode::BAD_REQUEST,
        Json(json!({ "error": msg.into() })),
    )
        .into_response()
}

fn not_found(msg: impl Into<String>) -> Response {
    (StatusCode::NOT_FOUND, Json(json!({ "error": msg.into() }))).into_response()
}

fn json_value(v: impl serde::Serialize) -> Json<Value> {
    Json(serde_json::to_value(v).unwrap_or_else(|e| json!({ "error": e.to_string() })))
}

/// Best-effort audit emit. Mirrors `requirements_routes::record_activity`
/// — failures are logged at WARN, never bubbled, because losing a
/// timeline row should not cost the comment write.
async fn record_comment_activity(project: &ProjectLayer, comment: &Comment, action: &str) {
    let Some(store) = project.activities.as_ref() else {
        return;
    };
    let body = json!({
        "comment_id": comment.id,
        "action": action,
        "parent_id": comment.parent_id,
        // Truncated preview for the UI; the full body lives on the
        // comment row and is fetched alongside the timeline.
        "preview": preview(&comment.body, 120),
    });
    let activity = Activity::new(
        comment.requirement_id.clone(),
        ActivityKind::Comment,
        comment.author.clone(),
        body,
    );
    if let Err(e) = store.append(&activity).await {
        warn!(
            requirement_id = %comment.requirement_id,
            comment_id = %comment.id,
            error = %e,
            "failed to record comment activity"
        );
    }
}

fn preview(body: &str, cap: usize) -> String {
    let trimmed = body.trim();
    if trimmed.chars().count() <= cap {
        return trimmed.to_string();
    }
    let mut s: String = trimmed.chars().take(cap).collect();
    s.push('…');
    s
}

// ----------------------- GET /v1/requirements/:id/comments --------------

async fn list_comments(State(project): State<ProjectLayer>, Path(id): Path<String>) -> Response {
    let store = match require_comment_store(&project) {
        Ok(s) => s,
        Err(r) => return r,
    };
    match store.list_for_requirement(&id).await {
        Ok(items) => json_value(json!({ "requirement_id": id, "items": items })).into_response(),
        Err(e) => internal_error(e),
    }
}

// ----------------------- POST /v1/requirements/:id/comments -------------

#[derive(Debug, Deserialize)]
struct PostCommentBody {
    body: String,
    /// `Some` to post a one-level reply. The store enforces the
    /// depth-1 invariant — we only forward.
    #[serde(default)]
    parent_id: Option<String>,
}

async fn post_comment(
    State(project): State<ProjectLayer>,
    Path(id): Path<String>,
    Json(payload): Json<PostCommentBody>,
) -> Response {
    let store = match require_comment_store(&project) {
        Ok(s) => s,
        Err(r) => return r,
    };
    let body = payload.body.trim();
    if body.is_empty() {
        return bad_request("comment body must not be empty");
    }
    if body.len() > MAX_COMMENT_BODY_BYTES {
        return bad_request(format!(
            "comment body exceeds {MAX_COMMENT_BODY_BYTES}-byte cap"
        ));
    }

    // v0 — every REST POST is a Human author. Real user identity
    // arrives in a future phase.
    let author = ActivityActor::Human;
    let comment = match payload.parent_id {
        Some(parent) if !parent.trim().is_empty() => {
            Comment::reply(id.clone(), author, body, parent.trim())
        }
        _ => Comment::new(id.clone(), author, body),
    };

    match store.create(&comment).await {
        Ok(()) => {
            record_comment_activity(&project, &comment, "posted").await;
            (StatusCode::CREATED, json_value(&comment)).into_response()
        }
        Err(e) => {
            // `parent_id` validation errors come from the store's
            // depth check — surface them as 400 rather than 500 so
            // the client knows the input was wrong.
            let msg = e.to_string();
            if msg.contains("depth 1") || msg.contains("not found") {
                bad_request(msg)
            } else {
                internal_error(e)
            }
        }
    }
}

// ----------------------- PATCH /v1/comments/:comment_id -----------------

/// `PATCH /v1/comments/:id?requirement_id=...` body.
#[derive(Debug, Deserialize)]
struct EditCommentBody {
    body: String,
}

#[derive(Debug, Deserialize)]
struct CommentIdQuery {
    /// Required — see `DeleteQuery` for the full rationale (the
    /// store partitions by requirement, so we need the parent id
    /// to dispatch).
    requirement_id: String,
}

async fn edit_comment(
    State(project): State<ProjectLayer>,
    Path(comment_id): Path<String>,
    axum::extract::Query(q): axum::extract::Query<CommentIdQuery>,
    Json(payload): Json<EditCommentBody>,
) -> Response {
    let store = match require_comment_store(&project) {
        Ok(s) => s,
        Err(r) => return r,
    };
    if q.requirement_id.trim().is_empty() {
        return bad_request("requirement_id query param is required");
    }
    let new_body = payload.body.trim();
    if new_body.is_empty() {
        return bad_request("comment body must not be empty");
    }
    if new_body.len() > MAX_COMMENT_BODY_BYTES {
        return bad_request(format!(
            "comment body exceeds {MAX_COMMENT_BODY_BYTES}-byte cap"
        ));
    }
    // Build a partial Comment carrying only the fields the trait
    // permits a caller to mutate; the store's `update` impl is
    // responsible for preserving `id`, `requirement_id`, `author`,
    // `parent_id`, `created_at` from the existing row. The other
    // fields here are placeholders the store ignores.
    let now = chrono::Utc::now().to_rfc3339();
    let patch = Comment {
        id: comment_id.clone(),
        requirement_id: q.requirement_id.clone(),
        author: ActivityActor::Human, // placeholder; preserved by store impl
        body: new_body.to_string(),
        parent_id: None, // placeholder; preserved by store impl
        created_at: now.clone(),
        updated_at: now,
    };

    match store.update(&patch).await {
        Ok(true) => {
            // Emit an Activity row so the timeline shows the edit.
            // Re-load the row first so the activity payload reflects
            // the actual author / parent_id of the comment, not the
            // placeholders we passed to `update`.
            let edited = match store.list_for_requirement(&q.requirement_id).await {
                Ok(rows) => rows.into_iter().find(|c| c.id == comment_id),
                Err(_) => None,
            };
            if let Some(c) = edited.as_ref() {
                record_comment_activity(&project, c, "edited").await;
                json_value(c).into_response()
            } else {
                // Edge case: row vanished between update and re-list
                // (concurrent delete). Surface success without the
                // activity emit; the timeline subscriber will still
                // see the prior `posted` row.
                json_value(json!({
                    "id": comment_id,
                    "requirement_id": q.requirement_id,
                    "body": new_body,
                }))
                .into_response()
            }
        }
        Ok(false) => not_found(format!("comment `{comment_id}` not found")),
        Err(e) => internal_error(e),
    }
}

// ----------------------- DELETE /v1/comments/:comment_id ----------------

#[derive(Debug, Deserialize)]
struct DeleteQuery {
    /// Required — the store partitions by requirement, so we need
    /// the parent id to dispatch the delete. Kept as a query param
    /// to keep the URL cacheable for GETs (parallel routes share
    /// `/v1/comments/:id`).
    requirement_id: String,
}

async fn delete_comment(
    State(project): State<ProjectLayer>,
    Path(comment_id): Path<String>,
    axum::extract::Query(q): axum::extract::Query<DeleteQuery>,
) -> Response {
    let store = match require_comment_store(&project) {
        Ok(s) => s,
        Err(r) => return r,
    };
    if q.requirement_id.trim().is_empty() {
        return bad_request("requirement_id query param is required");
    }
    match store.delete(&q.requirement_id, &comment_id).await {
        Ok(true) => {
            // We don't have the original Comment in hand here, so
            // emit a minimal activity row so the timeline shows the
            // delete. Author defaults to Human (v0 — see post handler).
            if let Some(act_store) = project.activities.as_ref() {
                let activity = Activity::new(
                    q.requirement_id.clone(),
                    ActivityKind::Comment,
                    ActivityActor::Human,
                    json!({
                        "comment_id": comment_id,
                        "action": "deleted",
                    }),
                );
                if let Err(e) = act_store.append(&activity).await {
                    warn!(
                        requirement_id = %q.requirement_id,
                        comment_id,
                        error = %e,
                        "failed to record comment-delete activity"
                    );
                }
            }
            json_value(json!({ "deleted": true })).into_response()
        }
        Ok(false) => not_found(format!("comment `{comment_id}` not found")),
        Err(e) => internal_error(e),
    }
}

File: packages/server/src/comments-routes.ts
// REST routes for Comment — per-requirement discussion threads.
//
// Ported from crates/harness-server/src/comments_routes.rs. Mounted only when
// `state.comments` is configured; every endpoint 503s otherwise so callers can
// distinguish "feature off" from "really broken".
//
// Endpoints:
//   GET    /v1/requirements/:id/comments  — list, oldest-first
//   POST   /v1/requirements/:id/comments  — post a top-level comment or reply
//   PATCH  /v1/comments/:comment_id       — edit body (?requirement_id=…)
//   DELETE /v1/comments/:comment_id       — delete (cascades to replies) (?requirement_id=…)
//
// Authoring policy (v0, mirrors Rust): every REST author is `Human`; real user
// identity arrives later. Body is trimmed + non-empty + length-capped. Every
// successful mutation also emits an `ActivityKind::Comment` row into
// `state.activities` (best-effort — a missing ActivityStore is silently skipped).
import type { FastifyInstance, FastifyReply } from "fastify";
import { errorText, type JsonValue } from "@jarvis/core";
import {
  newActivity,
  newComment,
  replyComment,
  type ActivityActor,
  type Comment,
  type CommentStore,
} from "@jarvis/project";
import type { AppState } from "./state.ts";

/**
 * Soft cap on a single comment body. Mirrors the Rust `MAX_COMMENT_BODY_BYTES`
 * (Linear: 32 KiB, GitHub: 64 KiB). Past this the model stops being useful as
 * discussion and the agent token budget suffers — push long content into the
 * description / a doc. Measured in UTF-8 bytes, matching Rust's `str::len()`.
 */
const MAX_COMMENT_BODY_BYTES = 32 * 1024;

const utf8 = new TextEncoder();

/** Return the comment store, or send a 503 and return undefined. */
function requireCommentStore(state: AppState, reply: FastifyReply): CommentStore | undefined {
  if (!state.comments) {
    reply.code(503).send({ error: "comment store not configured" });
    return undefined;
  }
  return state.comments;
}

/** Truncate a body for the activity preview, mirroring Rust's `preview`. */
function preview(body: string, cap: number): string {
  const trimmed = body.trim();
  const chars = [...trimmed];
  if (chars.length <= cap) return trimmed;
  return chars.slice(0, cap).join("") + "…";
}

/**
 * Best-effort audit emit. Mirrors `record_comment_activity` — failures are
 * swallowed (the timeline row is non-critical), and a missing ActivityStore is
 * a no-op.
 */
async function recordCommentActivity(state: AppState, comment: Comment, action: string): Promise<void> {
  if (!state.activities) return;
  const body: JsonValue = {
    comment_id: comment.id,
    action,
    parent_id: comment.parent_id ?? null,
    preview: preview(comment.body, 120),
  };
  const activity = newActivity(comment.requirement_id, "comment", comment.author, body);
  try {
    await state.activities.append(activity);
  } catch {
    /* losing a timeline row must not cost the comment write */
  }
}

export function registerCommentsRoutes(app: FastifyInstance, state: AppState): void {
  // ---- GET /v1/requirements/:id/comments ----
  app.get("/v1/requirements/:id/comments", async (req, reply) => {
    const store = requireCommentStore(state, reply);
    if (!store) return reply;
    const id = (req.params as { id: string }).id;
    try {
      const items = await store.listForRequirement(id);
      return reply.send({ requirement_id: id, items });
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
  });

  // ---- POST /v1/requirements/:id/comments ----
  app.post("/v1/requirements/:id/comments", async (req, reply) => {
    const store = requireCommentStore(state, reply);
    if (!store) return reply;
    const id = (req.params as { id: string }).id;
    const payload = (req.body ?? {}) as { body?: unknown; parent_id?: unknown };
    if (typeof payload.body !== "string") {
      return reply.code(400).send({ error: "comment body must not be empty" });
    }
    const body = payload.body.trim();
    if (body.length === 0) {
      return reply.code(400).send({ error: "comment body must not be empty" });
    }
    if (utf8.encode(body).length > MAX_COMMENT_BODY_BYTES) {
      return reply.code(400).send({ error: `comment body exceeds ${MAX_COMMENT_BODY_BYTES}-byte cap` });
    }

    // v0 — every REST POST is a Human author.
    const author: ActivityActor = { type: "human" };
    const parentRaw = typeof payload.parent_id === "string" ? payload.parent_id : undefined;
    const comment =
      parentRaw != null && parentRaw.trim().length > 0
        ? replyComment(id, author, body, parentRaw.trim())
        : newComment(id, author, body);

    try {
      await store.create(comment);
    } catch (e) {
      // `parent_id` validation errors come from the store's depth check —
      // surface them as 400 rather than 500 so the client knows the input was
      // wrong. (Mirrors Rust's substring match on "depth 1" / "not found".)
      const msg = errorText(e);
      if (msg.includes("depth 1") || msg.includes("not found")) {
        return reply.code(400).send({ error: msg });
      }
      return reply.code(500).send({ error: msg });
    }
    await recordCommentActivity(state, comment, "posted");
    return reply.code(201).send(comment);
  });

  // ---- PATCH /v1/comments/:comment_id?requirement_id=… ----
  app.patch("/v1/comments/:comment_id", async (req, reply) => {
    const store = requireCommentStore(state, reply);
    if (!store) return reply;
    const commentId = (req.params as { comment_id: string }).comment_id;
    const requirementId = (req.query as { requirement_id?: string }).requirement_id;
    if (typeof requirementId !== "string" || requirementId.trim().length === 0) {
      return reply.code(400).send({ error: "requirement_id query param is required" });
    }
    const payload = (req.body ?? {}) as { body?: unknown };
    if (typeof payload.body !== "string") {
      return reply.code(400).send({ error: "comment body must not be empty" });
    }
    const newBody = payload.body.trim();
    if (newBody.length === 0) {
      return reply.code(400).send({ error: "comment body must not be empty" });
    }
    if (utf8.encode(newBody).length > MAX_COMMENT_BODY_BYTES) {
      return reply.code(400).send({ error: `comment body exceeds ${MAX_COMMENT_BODY_BYTES}-byte cap` });
    }

    // Build a partial Comment carrying only the caller-mutable fields; the
    // store's `update` pins id / requirement_id / author / parent_id /
    // created_at from the existing row. The placeholders here are ignored.
    const now = new Date().toISOString();
    const patch: Comment = {
      id: commentId,
      requirement_id: requirementId,
      author: { type: "human" }, // placeholder; preserved by store impl
      body: newBody,
      created_at: now,
      updated_at: now,
    };

    let updated: boolean;
    try {
      updated = await store.update(patch);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
    if (!updated) {
      return reply.code(404).send({ error: `comment \`${commentId}\` not found` });
    }
    // Re-load so the activity payload + response reflect the real author /
    // parent_id, not the placeholders we passed to `update`.
    let edited: Comment | undefined;
    try {
      const rows = await store.listForRequirement(requirementId);
      edited = rows.find((c) => c.id === commentId);
    } catch {
      edited = undefined;
    }
    if (edited) {
      await recordCommentActivity(state, edited, "edited");
      return reply.send(edited);
    }
    // Edge case: row vanished between update and re-list (concurrent delete).
    return reply.send({ id: commentId, requirement_id: requirementId, body: newBody });
  });

  // ---- DELETE /v1/comments/:comment_id?requirement_id=… ----
  app.delete("/v1/comments/:comment_id", async (req, reply) => {
    const store = requireCommentStore(state, reply);
    if (!store) return reply;
    const commentId = (req.params as { comment_id: string }).comment_id;
    const requirementId = (req.query as { requirement_id?: string }).requirement_id;
    if (typeof requirementId !== "string" || requirementId.trim().length === 0) {
      return reply.code(400).send({ error: "requirement_id query param is required" });
    }
    let deleted: boolean;
    try {
      deleted = await store.delete(requirementId, commentId);
    } catch (e) {
      return reply.code(500).send({ error: errorText(e) });
    }
    if (!deleted) {
      return reply.code(404).send({ error: `comment \`${commentId}\` not found` });
    }
    // We don't have the original Comment in hand, so emit a minimal activity
    // row. Author defaults to Human (v0 — see post handler).
    if (state.activities) {
      const activity = newActivity(
        requirementId,
        "comment",
        { type: "human" },
        { comment_id: commentId, action: "deleted" },
      );
      try {
        await state.activities.append(activity);
      } catch {
        /* tolerate audit-emit failure */
      }
    }
    return reply.send({ deleted: true });
  });
}
